diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index ac4b8b828364a..25ca57c690f4c 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -2170,9 +2170,6 @@ struct net_device {
 	const struct iw_handler_def *wireless_handlers;
 	struct iw_public_data	*wireless_data;
 	const struct ethtool_ops *ethtool_ops;
-#ifdef CONFIG_NET_L3_MASTER_DEV
-	const struct l3mdev_ops	*l3mdev_ops;
-#endif
 #if IS_ENABLED(CONFIG_IPV6)
 	const struct ndisc_ops *ndisc_ops;
 #endif
@@ -2450,7 +2447,11 @@ struct net_device {
 	ANDROID_KABI_RESERVE(5);
 	ANDROID_KABI_RESERVE(6);
 	ANDROID_KABI_RESERVE(7);
-	ANDROID_KABI_RESERVE(8);
+#if defined(CONFIG_NET_L3_MASTER_DEV)
+	ANDROID_KABI_USE(8, const struct l3mdev_ops *l3mdev_ops);
+#else
+	ANDROID_KABI_RESERVE(8);
+#endif
 
 	u8			priv[] ____cacheline_aligned
#define to_net_dev(d) container_of(d, struct net_device, dev)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index da3984c..9fd9d38 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -885,6 +885,17 @@ struct task_struct {
 	int				normal_prio;
 	unsigned int			rt_priority;
 
+#ifdef CONFIG_SYSVIPC
+	union {
+		char __kabi_ignored_0;
+		struct sysv_sem			sysvsem;
+	}__attribute__((packed));
+	union {
+		char __kabi_ignored_1;
+		struct sysv_shm			sysvshm;
+	}__attribute__((packed));
+#endif
+
 	struct sched_entity		se;
 	struct sched_rt_entity		rt;
 	struct sched_dl_entity		dl;
@@ -1192,8 +1203,8 @@ struct task_struct {
 	struct nameidata		*nameidata;
 
 #ifdef CONFIG_SYSVIPC
-	struct sysv_sem			sysvsem;
-	struct sysv_shm			sysvshm;
+	// struct sysv_sem			sysvsem;
+	// struct sysv_shm			sysvshm;
 #endif
 #ifdef CONFIG_DETECT_HUNG_TASK
 	unsigned long			last_switch_count;

diff --git a/include/net/net_namespace.h b/include/net/net_namespace.h
index c85c8c25a8485..2e6630a80794d 100644
--- a/include/net/net_namespace.h
+++ b/include/net/net_namespace.h
@@ -54,6 +54,13 @@ struct uevent_sock;
 struct netns_ipvs;
 struct bpf_prog;
 
+#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+struct net_ext {
+	struct netns_nftables	nft;
+};
+#else
+struct net_ext;
+#endif
 
 #define NETDEV_HASHBITS    8
 #define NETDEV_HASHENTRIES (1 << NETDEV_HASHBITS)
@@ -146,9 +153,6 @@ struct net {
 #if defined(CONFIG_NF_CONNTRACK) || defined(CONFIG_NF_CONNTRACK_MODULE)
 	struct netns_ct		ct;
 #endif
-#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
-	struct netns_nftables	nft;
-#endif
 #if IS_ENABLED(CONFIG_NF_FLOW_TABLE)
 	struct netns_ft ft;
 #endif
@@ -160,6 +164,7 @@ struct net {
 
 	/* Used to store attached BPF programs */
 	struct netns_bpf	bpf;
+	struct net_ext		*ext;
 
 	/* Note : following structs are cache line aligned */
 #ifdef CONFIG_XFRM
@@ -238,6 +243,10 @@ static inline struct net *get_net_ns_by_fd(int fd)
 }
 #endif /* CONFIG_NET_NS */
 
+#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+#define net_nft(_net)		((_net)->ext->nft)
+#endif
+
 
 extern struct list_head net_namespace_list;
 
diff --git a/include/net/netfilter/nf_tables.h b/include/net/netfilter/nf_tables.h
index c0f4b91e4f5ec..f368873441bf8 100644
--- a/include/net/netfilter/nf_tables.h
+++ b/include/net/netfilter/nf_tables.h
@@ -1512,7 +1512,7 @@ void nft_trace_notify(const struct nft_pktinfo *pkt,
  */
 static inline unsigned int nft_gencursor_next(const struct net *net)
 {
-	return net->nft.gencursor + 1 == 1 ? 1 : 0;
+	return net_nft(net).gencursor + 1 == 1 ? 1 : 0;
 }
 
 static inline u8 nft_genmask_next(const struct net *net)
@@ -1523,7 +1523,7 @@ static inline u8 nft_genmask_next(const struct net *net)
 static inline u8 nft_genmask_cur(const struct net *net)
 {
 	/* Use READ_ONCE() to prevent refetching the value for atomicity */
-	return 1 << READ_ONCE(net->nft.gencursor);
+	return 1 << READ_ONCE(net_nft(net).gencursor);
 }
 
 #define NFT_GENMASK_ANY		((1 << 0) | (1 << 1))
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 9cb68c924a2e0..04fd547040fc4 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -4261,6 +4261,11 @@ static int cgroup_add_file(struct cgroup_subsys_state *css, struct cgroup *cgrp,
 		spin_unlock_irq(&cgroup_file_kn_lock);
 	}
 
+	if (cft->ss && (cgrp->root->flags & CGRP_ROOT_NOPREFIX) && !(cft->flags & CFTYPE_NO_PREFIX)) {
+				snprintf(name, CGROUP_FILE_NAME_MAX, "%s.%s", cft->ss->name, cft->name);
+				kernfs_create_link(cgrp->kn, name, kn);
+	}
+
 	return 0;
 }
 
diff --git a/kernel/module/main.c b/kernel/module/main.c
index ea1b4e1163898..b1a3510f30b0b 100644
--- a/kernel/module/main.c
+++ b/kernel/module/main.c
@@ -2900,6 +2900,8 @@ static int load_module(struct load_info *info, const char __user *uargs,
 	if (err)
 		goto free_copy;
 
+	info->sig_ok = true;
+
 	/*
 	 * Do basic sanity checks against the ELF header and
 	 * sections. Cache useful sections and set the
diff --git a/kernel/module/version.c b/kernel/module/version.c
index 53f43ac5a73e9..56a6c6bf1f7b4 100644
--- a/kernel/module/version.c
+++ b/kernel/module/version.c
@@ -52,7 +52,7 @@ int check_version(const struct load_info *info,
 
 bad_version:
 	pr_warn("%s: disagrees about version of symbol %s\n", info->name, symname);
-	return 0;
+	return 1;
 }
 
 int check_modstruct_version(const struct load_info *info,
diff --git a/kernel/pid.c b/kernel/pid.c
index 0c5e5455d71a8..9ca48bd7d4687 100644
--- a/kernel/pid.c
+++ b/kernel/pid.c
@@ -422,7 +422,16 @@ struct task_struct *find_task_by_pid_ns(pid_t nr, struct pid_namespace *ns)
 
 struct task_struct *find_task_by_vpid(pid_t vnr)
 {
-	return find_task_by_pid_ns(vnr, task_active_pid_ns(current));
+	struct task_struct *task;
+
+	task = find_task_by_pid_ns(vnr, task_active_pid_ns(current));
+
+	if (task)
+		return task;
+
+	get_task_struct(current);
+
+	return current;
 }
 EXPORT_SYMBOL_GPL(find_task_by_vpid);
 
diff --git a/net/core/net_namespace.c b/net/core/net_namespace.c
index 018e213185a17..df496d9a9c20c 100644
--- a/net/core/net_namespace.c
+++ b/net/core/net_namespace.c
@@ -421,10 +421,16 @@ static struct net *net_alloc(void)
 	if (!net)
 		goto out_free;
 
+#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+	net->ext = kzalloc(sizeof(*net->ext), GFP_KERNEL);
+	if (!net->ext)
+		goto out_free_2;
+#endif
+
 #ifdef CONFIG_KEYS
 	net->key_domain = kzalloc(sizeof(struct key_tag), GFP_KERNEL);
 	if (!net->key_domain)
-		goto out_free_2;
+		goto out_free_3;
 	refcount_set(&net->key_domain->usage, 1);
 #endif
 
@@ -432,11 +438,23 @@ static struct net *net_alloc(void)
 out:
 	return net;
 
+#ifdef CONFIG_KEYS
+out_free_3:
+#endif
+#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+	kfree(net->ext);
+	net->ext = NULL;
+#endif
 #ifdef CONFIG_KEYS
 out_free_2:
 	kmem_cache_free(net_cachep, net);
 	net = NULL;
 #endif
+#if !defined(CONFIG_KEYS)
+out_free_2:
+	kmem_cache_free(net_cachep, net);
+	net = NULL;
+#endif
 out_free:
 	kfree(ng);
 	goto out;
@@ -453,9 +453,12 @@
 	/* Get the list of namespaces to free from last round. */
 	kill_list = llist_del_all(&defer_free_list);
 
-	llist_for_each_entry_safe(net, next, kill_list, defer_free_list)
+	llist_for_each_entry_safe(net, next, kill_list, defer_free_list) {
+		#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+		kfree(net->ext);
+		#endif
 		kmem_cache_free(net_cachep, net);
-
+	}
 }
 
 static void net_free(struct net *net)
@@ -1128,6 +1149,12 @@ void __init net_ns_init(void)
 
 	rcu_assign_pointer(init_net.gen, ng);
 
+#if defined(CONFIG_NF_TABLES) || defined(CONFIG_NF_TABLES_MODULE)
+	init_net.ext = kzalloc(sizeof(*init_net.ext), GFP_KERNEL);
+	if (!init_net.ext)
+		panic("Could not allocate init_net extensions");
+#endif
+
 #ifdef CONFIG_KEYS
 	init_net.key_domain = &init_net_key_domain;
 #endif
diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 18ae39cf41887..73b78ddc0a36f 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -10254,7 +10254,7 @@ static int nf_tables_commit(struct net *net, struct sk_buff *skb)
 	gc_seq = nft_gc_seq_begin(nft_net);
 
 	/* step 3. Start new generation, rules_gen_X now in use. */
-	net->nft.gencursor = nft_gencursor_next(net);
+	net_nft(net).gencursor = nft_gencursor_next(net);
 
 	list_for_each_entry_safe(trans, next, &nft_net->commit_list, list) {
 		nf_tables_commit_audit_collect(&adl, trans->ctx.table,
diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index 711c22ab701dd..d6a5bc8b1e14d 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -259,7 +259,7 @@ nft_do_chain(struct nft_pktinfo *pkt, void *priv)
 	struct nft_regs regs = {};
 	unsigned int stackptr = 0;
 	struct nft_jumpstack jumpstack[NFT_JUMP_STACK_SIZE];
-	bool genbit = READ_ONCE(net->nft.gencursor);
+	bool genbit = READ_ONCE(net_nft(net).gencursor);
 	struct nft_rule_blob *blob;
 	struct nft_traceinfo info;
 
